### PR TITLE
Added OAuth2 authentication support

### PR DESCRIPTION
**Pull request by sfdrogojan**
Added OAuth2 authentication support

### CODE REVIEW

File: lib/marketingcloudsdk/client.rb [hunk 1]
```diff
@@ -121,11 +124,19 @@ def initialize(params={}, debug=false)
 			self.refresh_token = params['refresh_token'] if params['refresh_token']
 
 			self.wsdl = params["defaultwsdl"] if params["defaultwsdl"]
+
+			self.refresh
```
Comment (sfcbetiuc) [inline lib/marketingcloudsdk/client.rb]: Needed to support having only auth url in the params.

File: README.md [hunk 2]
```diff
@@ -6,6 +6,29 @@ ExactTarget Fuel SDK / SalesforceMarektingCloudSDK for Ruby
...
+            'request_token_url' => '<AUTH TENANT SPECIFIC ENDPOINT>',
+            'soap_endpoint' => '<SOAP TENANT SPECIFIC ENDPOINT>',
+            'base_api_url' => '<REST TENANT SPECIFIC ENDPOINT>',
+            'use_oAuth2_authentication' => true,
+            'account_id' => <TARGET_ACCOUNT_ID>,
+            'scope' => '<PERMISSION_LIST>'
+            }
+        })
+    ```
+
```
Comment (manivinesh) [inline README.md]: we also need to tell them that they can send the endpoint information through parameters of client class as well.

File: lib/marketingcloudsdk/client.rb [hunk 3]
```diff
@@ -121,11 +124,19 @@ def initialize(params={}, debug=false)
 			self.refresh_token = params['refresh_token'] if params['refresh_token']
 
 			self.wsdl = params["defaultwsdl"] if params["defaultwsdl"]
```
Comment (manivinesh) [inline lib/marketingcloudsdk/client.rb]: where is the code to read the OAuth2 options from parameters?

File: lib/marketingcloudsdk/client.rb [hunk 4]
```diff
@@ -105,6 +105,9 @@ def initialize(params={}, debug=false)
 				self.base_api_url = !(client_config["base_api_url"].to_s.strip.empty?) ? client_config["base_api_url"] : 'https://www.exacttargetapis.com'
 				self.request_token_url = !(client_config["request_token_url"].to_s.strip.empty?) ? client_config["request_token_url"] : 'https://auth.exacttargetapis.com/v1/requestToken'
```
Comment (manivinesh) [inline lib/marketingcloudsdk/client.rb]: if useOauth2 is true, we need to add a validation to see if there is Auth endpoint specified in config or params. If not, we need to throw exception

File: README.md (hunk 2 above)
Comment (sfdrogojan) [reply README.md]: The Ruby SDK supports only constructor parameters for the settings. The samples appear to use a config file, but that is a plain .rb file which declares a json object with all the settings. The file is than imported and that json object is mapped to the params constructor parameter.

File: lib/marketingcloudsdk/client.rb (hunk 3 above)
Comment (sfdrogojan) [reply lib/marketingcloudsdk/client.rb]: This is the constructor for the Client class:

def initialize(**params**={}, debug=false)
@refresh_mutex = Mutex.new
self.debug = debug
**client_config = params['client']**

All the settings are further retrieved from the **client_config** variable.

File: lib/marketingcloudsdk/client.rb (hunk 4 above)
Comment (sfdrogojan) [reply lib/marketingcloudsdk/client.rb]: Added validation.

**Pull request merged**